### PR TITLE
Four small correctness fixes in the HTTP server, Netty layer and form binding

### PR DESCRIPTION
**Pull request by graemerocher**
## Summary

Four small, unrelated fixes, one commit each.

- **`SystemFileBodyWriter`: honour single-byte ranges, answer 416 for unsatisfiable ones.**
  The range check required `first-pos < last-pos`, so `Range: bytes=0-0` - a valid one-byte
  range that download managers and media players use to probe range support - was treated as
  an invalid ranges-specifier and answered with a full `200` response. RFC 9110 section 14.1.2
  defines an int-range as valid when `first-pos <= last-pos`. A range whose first position is
  at or beyond the length of the file was handled the same way; RFC 9110 section 15.5.17
  requires `416 Range Not Satisfiable` with `Content-Range: bytes */<complete-length>` there.
  A zero-length representation keeps the previous behaviour, since a server may ignore a Range
  header field when the selected representation has no content.

- **`Http2AccessLogConnectionEncoder`: tolerate a stream that is no longer in the connection.**
  Both `writeHeaders` and `writeData` did `connection.stream(streamId).getProperty(...)`.
  `Http2Connection.stream(int)` returns `null` once the stream has been removed, so a write
  issued for a stream that is already gone threw a `NullPointerException` out of the decorator
  instead of reaching the delegate. Netty's own `DefaultHttp2ConnectionEncoder` handles that
  case by failing the write promise, which is what callers expect. The stream is now looked up
  first and the logging skipped when it is absent, so the write passes through and the promise
  carries the outcome.

- **`LoomCarrierGroup`: wake a parked carrier when a continuation is submitted externally.**
  A continuation submitted from a thread that is not part of the loop only triggered
  `backingHandler.wakeup()`, which interrupts a blocking IO operation. The carrier does not
  always wait inside one: whenever the IO virtual thread is parked, the carrier waits in
  `LockSupport.park()` in `Runner.run()` instead, and a selector wakeup does not reach it. The
  queued continuation was then only picked up once the IO virtual thread happened to be
  rescheduled for an unrelated reason. The carrier is now unparked alongside the wakeup, the
  same way the IO continuation is already scheduled. The unpark is cheap and idempotent: a
  carrier that is not parked simply skips its next park and re-checks its queues.

- **`HttpConverterRegistrar`: decode a form field with the charset declared by the part.**
  A multipart part may carry its own `Content-Type` with a `charset` parameter, and RFC 7578
  section 4.5 makes that charset the one that governs the value of the part. The parser keeps
  the part media type in the field metadata, but the `CompletedAttribute` converter decoded
  with the charset of the request. The charset declared by the part is now preferred, both for
  the direct conversion to `String` and for the conversion via `String` used for other target
  types; parts without a declared charset are unaffected.

## Testing

Every change ships with a regression test whose failure on the unmodified main source was
observed by checking out the base version of the changed file, running the test, and restoring
the fix.

- `FileTypeHandlerSpec` (`http-server-netty`): the byte-range data table gains `bytes=0-0` and
  `bytes=5-5` (206, `Content-Range: bytes 0-0/48`, one byte of body), and a new feature asserts
  `416` plus `Content-Range: bytes */48` for `bytes=9000-`, `bytes=9000-9100` and
  `bytes=<length>-`. The `bytes=9000-` row moved out of the "falls back to 200" table, since
  that case is exactly the unsatisfiable range that must now be a 416.
  Fail-before: 5 failures - the two 206 cases reported `200 == 206 false`, and the three 416
  cases reported `Expected exception of type 'HttpClientResponseException', but no exception was
  thrown`.

- `Http2AccessLogConnectionEncoderSpec` (new, `http-server-netty`): drives the encoder over a
  `DefaultHttp2Connection` and an `EmbeddedChannel` and writes headers and data for a stream id
  that was never registered.
  Fail-before: both features failed with `Expected no exception to be thrown, but got
  'java.lang.NullPointerException'`, caused by `Cannot invoke
  "io.netty.handler.codec.http2.Http2Stream.getProperty(...)" because the return value of
  "io.netty.handler.codec.http2.Http2Connection.stream(int)" is null`. After the fix the write
  promise completes as failed, as it does with Netty's undecorated encoder.

- `LoomCarrierSpec` (`http-server-netty`): a new feature parks the IO virtual thread of the loop
  on a `ReentrantLock` held by a plain thread, so the carrier waits outside of the IO operation,
  then submits a continuation from a thread that is not part of the loop and asserts that it
  runs while the IO thread is still parked. The controller gained a `/capture-scheduler`
  endpoint that records `EventLoopVirtualThreadScheduler.current()`.
  Fail-before: `Condition not satisfied: ran.await(10, TimeUnit.SECONDS)` - the continuation was
  never executed within the ten second budget (the feature took 11.2s). With the fix the same
  feature completes in well under a second.

- `FileUploadSpec` (`http-server-netty`): a new data-driven feature posts a hand-built multipart
  body whose single attribute declares `Content-Type: text/plain; charset=<charset>` and
  contains U+00E9, and asserts the value bound to `@Part String` round-trips. The `UTF-8`
  iteration is a control that passes before and after.
  Fail-before: the `ISO-8859-1` iteration failed with `1 difference (0% similarity)` between the
  replacement character and `é`.

Full test tasks and checkstyle for every module touched:

- `./gradlew :micronaut-http:test :micronaut-http-server:test :micronaut-http-netty:test
  --continue` - BUILD SUCCESSFUL, 1857 / 128 / 148 tests, no failures (3 skipped).
- `./gradlew :micronaut-http-server-netty:test` - 1162 tests, 19 skipped, 1 failure:
  `FiltersPropagatedContextSpec4 > test filter context propagation`, which failed with
  `HttpClientException: Client '/': Error occurred reading HTTP response: Connection reset`.
  That spec exercises filter context propagation and touches none of the code changed here
  (the loom carrier is only in play when `loom-carrier` is enabled, and the HTTP/2 access log
  encoder only when HTTP/2 access logging is on). Re-run on its own it passes:
  `./gradlew :micronaut-http-server-netty:test --tests
  "io.micronaut.http.server.netty.filters.FiltersPropagatedContextSpec4"` - BUILD SUCCESSFUL,
  1 test, 0 failures.
- `./gradlew :micronaut-http:checkstyleMain :micronaut-http-server:checkstyleMain
  :micronaut-http-netty:checkstyleMain :micronaut-http-server-netty:checkstyleMain
  :micronaut-http-server-netty:checkstyleTest` - BUILD SUCCESSFUL. The only reported violations
  are the two pre-existing ones, in files this branch does not touch:
  `NettyHttpServerConfiguration` FileLength and `NettyWebSocketSession` DesignForExtension.

## Out of scope

- `SystemFileBodyWriter` still answers `200` when the last position of the range is past the end
  of the file (`bytes=0-9000` on a 48 byte file). RFC 9110 section 14.1.2 says such a range
  should be clamped to the remainder of the representation and answered with a `206`. That is a
  separate behavioural change with its own existing test expectation, so it is left alone here.
- `parseRangeHeader` treats a suffix range (`bytes=-500`, meaning the last 500 bytes) as
  `bytes=0-500`. Also left alone.
- `Http2AccessLogManager.logHeaders` performs the same unguarded `connection.stream(streamId)`
  lookup, but on the request side where the stream has just been created; it is not part of this
  change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by graemerocher** [comment]
`FormFieldMetadata` is explicitly untrusted, but this now calls `MediaType.getCharset()` without handling an invalid client-supplied charset. That method delegates to `Charset.forName`, so a multipart part with `Content-Type: text/plain; charset=invalid` throws `UnsupportedCharsetException` out of conversion and produces HTTP 500.

I reproduced this with the new `/multipart/attribute` endpoint: the response was 500 and the stack trace points to `HttpConverterRegistrar.declaredCharset`. This also conflicts with the existing `HttpUtilSpec` behavior that invalid charsets fall back to the default. Please catch `IllegalCharsetNameException` / `UnsupportedCharsetException` here and either fall back to `context.getCharset()` (by returning no declared charset) or reject the conversion as a client-input error rather than surfacing a server error.

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/LoomCarrierSpec.groovy [hunk 1]
```diff
@@ -94,9 +98,74 @@ class LoomCarrierSpec extends Specification {
...
+            lock.lock()
+            lock.unlock()
+            ioResumed.countDown()
+        })
+        long deadline = System.currentTimeMillis() + 10_000
+        while (!lock.hasQueuedThreads() && System.currentTimeMillis() < deadline) {
+            Thread.sleep(10)
+        }
+        // let the carrier finish the loop iteration and settle
+        Thread.sleep(500)
```
[the remark this answers is not in the document]
Comment (graemerocher) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/LoomCarrierSpec.groovy]: Fixed in f5c80b7b50. The await result is now asserted.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/LoomCarrierSpec.groovy [hunk 2]
```diff
@@ -94,9 +98,74 @@ class LoomCarrierSpec extends Specification {
...
+        }, "loom-carrier-spec-holder")
+
+        when: "the io thread of the loop parks on a lock, so its carrier waits outside of the io operation"
+        holder.start()
+        held.await(10, TimeUnit.SECONDS)
+        scheduler.eventLoop().execute({
+            lock.lock()
+            lock.unlock()
+            ioResumed.countDown()
+        })
```
[the remark this answers is not in the document]
Comment (graemerocher) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/LoomCarrierSpec.groovy:141]: Fixed in f5c80b7b50. The poll loop and the 500ms "let the carrier settle" sleep are replaced by a single `PollingConditions` wait on `lock.hasQueuedThreads() && carrier.state == Thread.State.WAITING`. That second condition is the real edge the sleep was estimating: once the IO thread queues on the lock it unmounts, the carrier finishes its iteration and parks in `LockSupport.park()`, staying `WAITING` until woken. Re-verified the fail-before still holds with the production change reverted, and the spec went from over 11s to 6.5s.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/accesslog/Http2AccessLogConnectionEncoderSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    void "writeData for a stream that is not in the connection fails the promise"() {
+        given:
+        ChannelHandlerContext ctx = channel.pipeline().firstContext()
+        ByteBuf data = Unpooled.copiedBuffer("hello", StandardCharsets.UTF_8)
+        ChannelPromise promise = channel.newPromise()
+
+        when:
+        encoder.writeData(ctx, 3, data, 0, true, promise)
```
[the remark this answers is not in the document]
Comment (graemerocher) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/handler/accesslog/Http2AccessLogConnectionEncoderSpec.groovy:47]: Fixed in f5c80b7b50. A `cleanup:` block releases `data` if it still holds references; the `refCnt() == 0` assertion on the success path is unchanged.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by graemerocher** [comment]
Confirmed and fixed in 77bd2f9c7e. Reproduced exactly as you described: `charset=invalid` gives a 500 with `UnsupportedCharsetException` at `HttpConverterRegistrar.declaredCharset`, and `charset=8859_1!` gives `IllegalCharsetNameException` from the other exit of `Charset.forName`.

I took the fall-back option rather than rejecting as client input, because that is what the codebase already does: `HttpUtil.resolveCharset` and `HttpHeadersUtil.parseCharacterEncoding` both swallow an unresolvable charset and return the default, and `HttpUtilSpec` pins that. Rejecting would make a part stricter than the message containing it. `declaredCharset` now catches both exceptions and returns null, which is the existing "part declares no charset" path.

The regression test is data-driven over both `invalid` and `8859_1!` so each catch branch executes, asserting 200 and that U+00E9 round-trips via the request charset.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by dstepanov — APPROVED** [review]
LGTM. The HTTP/2 access-log null guard, the Loom carrier unpark and the part charset decoding are all correct, and the Range fix works for `bytes=0-0` and `bytes=N-`. Some edge cases in [`SystemFileBodyWriter`](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java#L104-L135), none of them likely in practice:

1. **`bytes=-0` now returns 206 with the first byte (it was 200).** [`parseRangeHeader`](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java#L187) maps an empty first-pos to 0, so `-0` becomes `0-0`, which the new [`<=` check](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java#L124) accepts. RFC 9110 §14.1.1 says a zero-length suffix is unsatisfiable. The same parser still reads `bytes=-5` as `0-5`, which you already called out of scope.
2. **Some invalid headers get 416 instead of being ignored.**
   - The [past-the-end check](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java#L114) runs before the `firstPos <= lastPos` validity check. On a 48-byte file, `bytes=9000-10` gets 416 while `bytes=10-5` gets 200. Both are invalid, and §14.1.2 says to ignore the header.
   - The unit check is [`startsWith("bytes")`](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java#L108), so `bytesX=9000-` now gets 416 too.
3. **`If-Range` is still ignored.** A download resumed with a stale `If-Range` against a file that has since shrunk now gets 416. §13.1.5 wants 200 with the full representation.
4. **The empty 416 is cacheable.** It carries the file's `Cache-Control: public, max-age=…` from [`handleIfModifiedAndHeaders`](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java#L94), so a CDN could cache the error. Low confidence.
5. **Part charset is honoured in only one place.**
   - `@Part String` now uses the part's charset, but `SharedHttpConvertersRegistrar`'s `CompletedAttribute` converter and `FormRouteCompleter.mapForGetBody` (`@Body` Map/POJO) still use the request charset. The same part can decode two different ways.
   - [Any JVM charset](https://github.com/micronaut-projects/micronaut-core/blob/f5c80b7b50ae7ae13a3cb611a678e0ed5f5eb1dd/http/src/main/java/io/micronaut/http/converters/HttpConverterRegistrar.java#L176-L186) a client names is accepted (`IBM037`, `UTF-16`). That lets content decode differently from what a WAF saw; ASP.NET had this exact bypass. Browsers never send a per-part charset, so restricting it to ASCII-compatible charsets may be worth considering.

Missing tests: `bytes=-0`, `bytes=-N`, `bytes=9000-10`, HEAD with Range, a zero-length file, `@Body` binding with a part charset.

**Comment by graemerocher** [comment]
## Codex review (gpt-5.6-sol, reasoning effort high)

An independent review run with the OpenAI Codex CLI against this PR's diff, posted verbatim, followed by what was done with each finding. Fixes are prepared locally on this branch and will be pushed when the current CI hold is lifted.

---

The primary cases are improved, but declared multipart charsets remain ineffective for wrapped string targets, and the range change incorrectly serves a byte for the valid-but-unsatisfiable `bytes=-0` suffix range.


- [P2] Decode wrapped strings with the part charset — http/src/main/java/io/micronaut/http/converters/HttpConverterRegistrar.java:140-143
  When a controller declares a wrapped target such as `@Part Optional<String> value`, `targetType` is `Optional.class`, so this new fast path is skipped. The subsequent `ReadBuffer`-to-`Optional` conversion recursively decodes the inner string using `context.getCharset()`, returns a non-empty result, and bypasses the declared-charset fallback, causing parts such as ISO-8859-1 attributes to bind as mojibake. Apply the part charset before converting container targets as well.

- [P2] Reject zero-length suffix ranges instead of serving a byte — http-server/src/main/java/io/micronaut/http/server/body/SystemFileBodyWriter.java:124-126
  For `Range: bytes=-0` on any non-empty file, `parseRangeHeader` produces `(0, 0)`, and the relaxed `<=` check now returns a 206 containing the first byte with `Content-Range: bytes 0-0/...`. RFC 9110 defines a suffix range of length zero as unsatisfiable, so this case must not be treated like the explicit one-byte range `0-0`; distinguish suffix ranges or reject `-0` before this check.

---

**Disposition:** Both verified. Fixed locally: `3a5f48e53f` implements suffix ranges per RFC 9110 (the parser never had them, so `bytes=-500` was served as the *first* 501 bytes) and answers 416 for `bytes=-0`; `3f0ae040f7` decodes every non-binary target with the declared charset, so `@Part Optional<String>` no longer binds ISO-8859-1 as mojibake. Both with spec rows that fail on the current head.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
